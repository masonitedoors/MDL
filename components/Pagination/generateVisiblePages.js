export function generateVisiblePages(currentPage, lastPage) {
  const delta = 2
  const range = []

  for (
    let i = Math.max(2, currentPage - delta);
    i <= Math.min(lastPage - 1, currentPage + delta);
    i += 1
  ) {
    range.push(i)
  }

  if (currentPage - delta > 2) {
    range.unshift('...')
  }

  if (currentPage + delta < lastPage - 1) {
    range.push('...')
  }

  range.unshift(1)

  if (lastPage !== 1) range.push(lastPage)

  return range
}
