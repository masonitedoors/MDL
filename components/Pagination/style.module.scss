@import "styles/index";

.pagination {
  border-top: 1px solid $color-night;
  border-bottom: 1px solid $color-night;
  padding: 8px 0;
  position: relative;
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;

  &__info {
    margin: 0;
    padding: 0;
    font-size: 12px;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
  }

  &__info-viewing-label {
    text-transform: uppercase;
    font-weight: 700;

    &::after {
      content: " ";
    }
  }

  &__nav {
    margin: 0;
    padding: 0;
    display: flex;
    list-style: none;
  }

  &__nav-item {
    margin: 0;
    padding: 0;
    display: flex;
    justify-content: space-between;
    align-items: center;

    &--active > * {
      text-shadow: 0 0 1px $color-night;
    }
  }

  &__nav-item-prev {
    margin-right: $space-xs;
  }

  &__nav-item-next {
    margin-left: $space-xs;
  }

  &__nav-link {
    margin: 0;
    padding: 0 0.5em;
    background-color: transparent;
    border-radius: 2px;
    border: 0;
    cursor: pointer;
    font-size: 15px;
    line-height: inherit;
    text-align: center;
    text-decoration: none;
    transition: background-color 0.2s;
    user-select: none;
    -webkit-appearance: none;

    &:focus {
      z-index: 1; // Ensure focus outline stays on top when user hovers an adjacent link.
    }
  }

  &__nav-ellipsis {
    margin: 0;
    padding: 0 0.5em;
    font-size: 15px;
    line-height: inherit;
    user-select: none;
  }
}
