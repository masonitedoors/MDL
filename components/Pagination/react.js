import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames/bind'
import { React as Button } from 'components/Button'
import { generateVisiblePages } from './generateVisiblePages'

import styles from './style.module.scss'

const cx = classNames.bind(styles)

const Pagination = ({
  page,
  totalItems,
  itemsPerPage,
  onChange,
  prevLabel,
  nextLabel,
  viewingLabel,
}) => {
  const totalPages = Math.ceil(totalItems / itemsPerPage)
  const pageNumbers = [...Array(totalPages).keys()].map(n => n + 1)

  const firstPageNumber = 1
  const lastPageNumber = pageNumbers[pageNumbers.length - 1]
  const prevPageNumber = page - 1
  const nextPageNumber = page + 1

  const viewingMin = page * itemsPerPage - (itemsPerPage - 1)
  const viewingMax = page === lastPageNumber ? totalItems : page * itemsPerPage
  const viewingRange = `${viewingMin}-${viewingMax} of ${totalItems}`

  const visiblePageNumbers = generateVisiblePages(page, totalPages)

  return (
    <div className={cx('pagination')}>
      <div className={cx('pagination__info')}>
        <span className={cx('pagination__info-viewing-label')}>{viewingLabel}</span>
        {viewingRange}
      </div>
      <nav role="navigation" aria-label="Pagination Navigation">
        <ul className={cx('pagination__nav')}>
          <li className={cx(['pagination__nav-item', 'pagination__nav-item-prev'])}>
            <Button
              size="small"
              aria-label={`Page ${prevPageNumber}`}
              disabled={page === firstPageNumber}
              onClick={() => onChange(prevPageNumber)}
              uppercase={true}
              bold={true}
            >
              {prevLabel}
            </Button>
          </li>

          {visiblePageNumbers.map((pageNumber, i) => {
            const key = `${i}_${pageNumber}`
            return (
              <li
                key={key}
                className={cx([
                  'pagination__nav-item',
                  page === pageNumber && 'pagination__nav-item--active',
                ])}
              >
                {pageNumber === '...' ? (
                  <span className={cx(['pagination__nav-ellipsis'])}>...</span>
                ) : (
                  <button
                    type="button"
                    className={cx('pagination__nav-link')}
                    aria-label={`Page ${pageNumber}`}
                    onClick={() => onChange(pageNumber)}
                  >
                    {pageNumber}
                  </button>
                )}
              </li>
            )
          })}

          <li className={cx(['pagination__nav-item', 'pagination__nav-item-next'])}>
            <Button
              size="small"
              aria-label={`Page ${nextPageNumber}`}
              disabled={page === lastPageNumber}
              onClick={() => onChange(nextPageNumber)}
              uppercase={true}
              bold={true}
            >
              {nextLabel}
            </Button>
          </li>
        </ul>
      </nav>
    </div>
  )
}

Pagination.propTypes = {
  page: PropTypes.number.isRequired,
  totalItems: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
  itemsPerPage: PropTypes.number,
  prevLabel: PropTypes.string,
  nextLabel: PropTypes.string,
  viewingLabel: PropTypes.string,
}

Pagination.defaultProps = {
  itemsPerPage: 10,
  prevLabel: 'Prev',
  nextLabel: 'Next',
  viewingLabel: 'Viewing',
}

export default Pagination
