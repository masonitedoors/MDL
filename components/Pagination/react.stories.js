import React from 'react'
import { storiesOf } from '@storybook/react'
import { withState } from '@dump247/storybook-state'
import { withKnobs, number } from '@storybook/addon-knobs'
import Pagination from './react'

storiesOf('Pagination', module)
  .addDecorator(withKnobs)
  .add(
    'default',
    withState({ page: 1 })(({ store }) => {
      const itemsPerPage = number('Number of items per page', 10)
      const totalItems = number('Total number of items', 103)

      return (
        <Pagination
          page={store.state.page}
          totalItems={totalItems}
          itemsPerPage={itemsPerPage}
          onChange={page => store.set({ page })}
        />
      )
    }),
  )
