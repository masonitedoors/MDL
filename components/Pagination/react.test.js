import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import Pagination from './react'

describe('Pagination', () => {
  let wrapper
  let props

  beforeEach(() => {
    props = {
      page: 4,
      totalItems: 103,
      itemsPerPage: 10,
      onChange: jest.fn(),
    }
    wrapper = shallow(<Pagination {...props} />)
  })

  it('Should render the component without crashing', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should display the correct number of visible pages based on total items and items per page', () => {
    expect(wrapper.find('.pagination__nav-item').length).toBe(10)
  })

  it('Should return the correct page number when a user clicks a specific page', () => {
    const buttonPage3 = wrapper.find('button[aria-label="Page 3"]')
    buttonPage3.simulate('click')
    expect(props.onChange).toHaveBeenCalledTimes(1)
    expect(props.onChange).toHaveBeenCalledWith(3)
  })

  describe('When the first page is active', () => {
    let prevButton
    let nextButton

    beforeEach(() => {
      props = {
        ...props,
        page: 1,
      }
      wrapper = shallow(<Pagination {...props} />)
      prevButton = wrapper.find('.pagination__nav-item-prev Button')
      nextButton = wrapper.find('.pagination__nav-item-next Button')
    })

    it('Should disable the `prev` page button', () => {
      expect(prevButton.props().disabled).toBe(true)
    })

    it('Should enable the `next` page button', () => {
      expect(nextButton.props().disabled).toBe(false)
    })
  })

  describe('When the last page is active', () => {
    let prevButton
    let nextButton

    beforeEach(() => {
      props = {
        ...props,
        page: 11,
      }
      wrapper = shallow(<Pagination {...props} />)
      prevButton = wrapper.find('.pagination__nav-item-prev Button')
      nextButton = wrapper.find('.pagination__nav-item-next Button')
    })

    it('Should enable the `prev` page button', () => {
      expect(prevButton.props().disabled).toBe(false)
    })

    it('Should disable the `next` page button', () => {
      expect(nextButton.props().disabled).toBe(true)
    })
  })
})
