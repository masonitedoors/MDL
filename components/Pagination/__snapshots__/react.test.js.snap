// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Should render the component without crashing 1`] = `
<div
  className="pagination"
>
  <div
    className="pagination__info"
  >
    <span
      className="pagination__info-viewing-label"
    >
      Viewing
    </span>
    31-40 of 103
  </div>
  <nav
    aria-label="Pagination Navigation"
    role="navigation"
  >
    <ul
      className="pagination__nav"
    >
      <li
        className="pagination__nav-item pagination__nav-item-prev"
      >
        <Button
          aria-label="Page 3"
          disabled={false}
          fullWidth={false}
          onClick={[Function]}
          size="small"
          uppercase={false}
          variant="default"
        >
          Prev
        </Button>
      </li>
      <li
        className="pagination__nav-item"
        key="0_1"
      >
        <button
          aria-label="Page 1"
          className="pagination__nav-link"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
      </li>
      <li
        className="pagination__nav-item"
        key="1_2"
      >
        <button
          aria-label="Page 2"
          className="pagination__nav-link"
          onClick={[Function]}
          type="button"
        >
          2
        </button>
      </li>
      <li
        className="pagination__nav-item"
        key="2_3"
      >
        <button
          aria-label="Page 3"
          className="pagination__nav-link"
          onClick={[Function]}
          type="button"
        >
          3
        </button>
      </li>
      <li
        className="pagination__nav-item pagination__nav-item--active"
        key="3_4"
      >
        <button
          aria-label="Page 4"
          className="pagination__nav-link"
          onClick={[Function]}
          type="button"
        >
          4
        </button>
      </li>
      <li
        className="pagination__nav-item"
        key="4_5"
      >
        <button
          aria-label="Page 5"
          className="pagination__nav-link"
          onClick={[Function]}
          type="button"
        >
          5
        </button>
      </li>
      <li
        className="pagination__nav-item"
        key="5_6"
      >
        <button
          aria-label="Page 6"
          className="pagination__nav-link"
          onClick={[Function]}
          type="button"
        >
          6
        </button>
      </li>
      <li
        className="pagination__nav-item"
        key="6_..."
      >
        <span
          className="pagination__nav-ellipsis"
        >
          ...
        </span>
      </li>
      <li
        className="pagination__nav-item"
        key="7_11"
      >
        <button
          aria-label="Page 11"
          className="pagination__nav-link"
          onClick={[Function]}
          type="button"
        >
          11
        </button>
      </li>
      <li
        className="pagination__nav-item pagination__nav-item-next"
      >
        <Button
          aria-label="Page 5"
          disabled={false}
          fullWidth={false}
          onClick={[Function]}
          size="small"
          uppercase={false}
          variant="default"
        >
          Next
        </Button>
      </li>
    </ul>
  </nav>
</div>
`;
